Codegen: More AST nodes

bjodah opened the pull request:
Goals of this PR:

- [x] Introduces more AST nodes (While, FunctionDefinition etc.).
- [x] C-code printing of new nodes
- [x] Internal utility package (`.utilities._compilation`) for compiling C code for "on-the-fly" for testing (based on [pycompilation](https://github.com/bjodah/pycompilation)).
- [x] `.codegen.algorithms` module for collecting functions which generate ast trees for common algorithms (right now only Newton's method)
- [x] Fortran & Python printing of new nodes
- [x] Testing of printing of new nodes

TODO:
Using attributes with ast nodes should also support the attributes having arguments, e.g.:
- [x] change `attrs` from `FiniteSet` to `Tuple`
- [x] ~~see if I can make things work without `none`~~ it's useful and needed to represent `None` in pycode.
- [x] revert change of `CodeBlock` signature
- [x] dimension (Fortran)
- [x] implement pretty printing of AST.
- [x] Store **all** optional arguments as `attrs` (which is stored last in `.args`)
    - [x] ~~Change `.type` to be an attribute~~ Untyped variables will have `.type == untyped`.
    - [x] Provide sensible defaults in `__new__` or other classmethod (i.e. wrap arguments and put them in attrs when needed)

bjodah committed:
New classes in .codegen.ast

While, Scope, PrintStatement, FunctionPrototype, FunctionDefinition, FunctionCall

bjodah committed:
Fix tests

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Add submodule .utilities._compilation for internal use

bjodah committed:
Add .codegen.algorithms with Newton's method

bjodah commented:
@moorepants just FYI, I've ported most parts of pycompilation into `.utilities._compilation` in this PR. (I needed to test the C code generated by `.codegen.algorithms`)

bjodah committed:
Add .utilities._compilation.util.capture_stdout_stderr

bjodah committed:
Tweak .codegen.ast.Token compatibility with Basic

bjodah committed:
Add new ast nodes to test_ccode.py

bjodah committed:
Handle attributes through .codegen.ast.Node

asmeurer commented on sympy/codegen/algorithms.py:
```diff
@@ -0,0 +1,93 @@
+    expr : expression
+    wrt : Symbol
+        With respect to, i.e. what is the variable
+
```
Add a doctest here

asmeurer commented on sympy/codegen/ast.py:
```diff
@@ -452,15 +465,18 @@ def body(self):
+        args = args + tuple([kwargs.pop(k) for k in cls.__slots__[len(args):]])
+        if kwargs:
+            raise ValueError("Unknown kwargs: %s" % kwargs)
         obj = Basic.__new__(cls)
```
I was playing with this and it's a bit odd that the typical ast node is completely atomic from a SymPy point of view, because nothing is stored in .args. I think .args should be used whereever possible to store the ast node expression tree.

bjodah replied:
ok, it's usually `str` or `None` that causes problems, (e.g. `Variable.type` can now be `None`), representing type through attributes would change that though.

asmeurer replied:
Well those specific objects would have empty args (the ones that emulate Symbol). The other nodes that contain other nodes should do so with .args.

bjodah committed:
Refactor Nodes in .codegen.ast

bjodah committed:
More .codegen.ast nodes

bjodah commented on sympy/codegen/ast.py:
```diff
@@ -481,15 +524,73 @@ def __hash__(self):
+        return super(Token, self).__hash__()
+
+
+none = NoneToken()
```
Using this "trick" (together with `class String`) allowed me to use `Basic.args` throughout―now SymPy's `replace` works as expected. @asmeurer I hope this is acceptable (it only lives in .codegen.ast after all)?

bjodah replied:
Turns out this is also useful for representing the value `None` when printing Python code using `pycode`.

asmeurer commented on sympy/codegen/ast.py:
```diff
@@ -481,15 +524,73 @@ def __hash__(self):
+    @classmethod
+    def _construct_attrs(cls, arg):
+        if arg is None:
+            return Tuple()
```
Why can't the default just be `Tuple()`?

bjodah replied:
You're right (I think I missed this one when introducing `none`)

bjodah committed:
Better str/repr printing of .codegen.ast nodes

bjodah committed:
Initial support for pretty printing codegen ASTs

bjodah committed:
Revert CodeBlock.body -> CodeBlock.args, fix indented printing

bjodah committed:
Update algorithms since update of CodeBlock

bjodah committed:
Work toward enabling fortran code printers for new ast nodes

bjodah committed:
Support for new AST nodes in Python & Fortran

bjodah committed:
Intermediate commit

asmeurer commented on sympy/codegen/fnodes.py:
```diff
@@ -123,37 +174,120 @@ def allocated(array):
+
+
+class dsign(FFunction):
+    """ Fortran sign intrinsic with for double precision arguments. """
```
Remove "with"

asmeurer commented on sympy/utilities/_compilation/__init__.py:
```diff
@@ -15,4 +15,63 @@
+from .compilation import compile_link_import_strings, compile_run_strings
+
+
+def has_fortran():
```
It's better to have these functions in the submodule than in `__init__.py`

bjodah committed:
Fix test failures for temporary commit

bjodah committed:
Test fixes, Variable.as_Declaration

bjodah committed:
Fix test failures

bjodah committed:
Enhanced documentation (and minor tweaks)

bjodah committed:
Enhanced documentation, some additions to .codegen.fnodes

bjodah committed:
Fix missing classes to test_args, fix doc warnings

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah commented:
There was a PR merged which caused a merge conflict with the PR. I merged with master and resolved the conflict. But now Travis is complaining about the use a an invalid escape sequence.

Is there a way to locate where this offending escape sequence is?

bjodah committed:
Fix merge

bjodah commented:
It was actually me missing a one conflict. I found it by running this script (stops at the first file which fails to compile):
```bash
#!/bin/bash
for f in $(find sympy -name "*.py"); do
    output=$(python3 -We:invalid -m compileall -f $f)
    if [ $? -ne 0 ]; then
       echo $output
       break;
    fi
done
```

smichr commented on sympy/codegen/__init__.py:
```diff
@@ -1 +1,18 @@
-from .ast import Assignment, aug_assign, CodeBlock, For
+""" The ``sympy.codegen`` module contains classes and functions for building
+abstract syntax trees of algorithms. This trees may then be printed by the
```
This -> These

smichr commented on sympy/codegen/algorithms.py:
```diff
@@ -0,0 +1,125 @@
+        Will be a ``Dummy`` if ``None``.
+
+    Examples
+    --------
```
-------- -> ========
and a blank line after it

smichr replied:
here and throughout (unless rst is different than the usual requirements)

smichr commented on sympy/codegen/algorithms.py:
```diff
@@ -0,0 +1,125 @@
+
+    References
+    ==========
+    https://en.wikipedia.org/wiki/Newton%27s_method
```
not sure if a blank line is needed here -- see rendering and/or check other files

smichr commented on sympy/codegen/algorithms.py:
```diff
@@ -0,0 +1,125 @@
+def _symbol_of(arg):
+    if isinstance(arg, Declaration):
+        arg = arg.variable.symbol
+    if isinstance(arg, Variable):
```
if -> elif?

smichr commented on sympy/codegen/algorithms.py:
```diff
@@ -0,0 +1,125 @@
+    """ Generates an AST for a function implementing the Newton-Raphson method.
+
+    Parameters
+    ----------
```
----------- -> ==========
and blank line following

smichr commented on sympy/codegen/cfunctions.py:
```diff
@@ -1,5 +1,6 @@
 """
-Functions with corresponding implementations in C.
+This module contains SymPy functions mathcin corresponding special math functions in the
```
corresponding -> corresponding to

smichr commented on sympy/codegen/cxxnodes.py:
```diff
@@ -0,0 +1,14 @@
+"""
+AST nodes specific for C++.
```
specific for -> specific to

smichr commented on sympy/codegen/cnodes.py:
```diff
@@ -0,0 +1,102 @@
+
+
+class PostDecrement(Basic):
+    """ Represents the post-decrement oparator """
```
oparator -> operator

smichr replied:
and throughout

smichr commented on sympy/codegen/fnodes.py:
```diff
@@ -0,0 +1,616 @@
+pure = Attribute('pure')
+elemental = Attribute('elemental')  # (all elemental procedures are also pure)
+
+intent_in = Attribute('intent_in')
```
Did you intend to write 'indent' instead of 'intent'?

bjodah replied:
Sorry about all the misspellings, in this case though I do mean "intent" (it's a Fortran concept: `intent(in)`, `intent(out)`, `intent(inout)`)

bjodah commented on sympy/codegen/fnodes.py:
```diff
@@ -0,0 +1,616 @@
+    _construct_only = staticmethod(lambda args: Tuple(*[arg if isinstance(arg, use_rename) else _name(arg) for arg in args]))
+
+
+class Module(Token):
```
Thanks! Do you want me to address your comments now, or do you want me to wait (if you want to review more later?)

smichr replied:
Go ahead and do some now. I'll try to keep going over the next week.

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Address review from @smichr

bjodah committed:
Drop _print_Integer from mpmath printer

bjodah commented:
I know this is long overdue but I took the time to merge in master and fix conflicts & address @smichr review so far.

bjodah committed:
Minor improvements to AST nodes, added test_applications.

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Support lt, le, ge, gt on .codegen.ast.Variable

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Settle for only passing kwargs generally in _print_* methods

stewpend0us committed:
wrote a failing test for CodeBlock.cse(). I could have the assertion wrong but the .cse() method actually fails to run. I tracked it down to cse_main.py -> _rebuild(expr): -> if iterable(expr): when expr is a tuple of (more than 1) Matrix the _rebuild method doesn't correctly rebuild the elements of the tuple. Matrix class is expecting (Matrix([..]),) as it's arguments but it's getting Matrix([..]) (missing the sinlge element tuple) I don't konw if that's a problem with the matrix stuff or the cse stuff....but either way :(

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Added doctest to .codegen.algorithms.newtons_method_function

bjodah committed:
Make test for gh-14118 xfail for now

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

stewpend0us committed:
couple of comments that would have helped me. Also some tests in the printer for Element

bjodah committed:
Merge pull request #4 from stewpend0us/codegen-ast-more-nodes

couple of comments that would have helped me

bjodah committed:
fix codegen.rst

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Drop stewpend0us tests for now

bjodah committed:
Update codegen.rst

stewpend0us committed:
some low hanging fruit...

stewpend0us committed:
added missing test

bjodah committed:
Merge pull request #7 from stewpend0us/codegen-ast-more-nodes

some low hanging fruit...

bjodah committed:
Remove spurious __new__ (fixes bjodah/sympy#8)

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

cym1 committed:
Update beam.py

The previous version doesn't allow applying reaction moments.

Example, a beam has length l and fixed supports at both ends. A 1000N load is applied at the center.

E = Symbol('E')
I = Symbol('I')
l = Symbol('l', positive=True)
R1, R2 = symbols('R1, R2')
M1, M2 = symbols('M1, M2')

b = Beam(l, E, I)
b.bc_deflection = [(0, 0),(l, 0)]
b.bc_slope = [(0, 0),(l, 0)]

b.apply_load(R1, 0, -1)
b.apply_load(M1, 0, -2)
b.apply_load(R2, l, -1)
b.apply_load(M2, l, -2)
b.apply_load(-1000, 0.5*l, -1)

Method solve_for_reaction_loads will not solve M1 and M2. A new method solve() is proposed here to find M1, M2.

A new method extrema() also proposed to find the maximum deflection points

cym1 committed:
Update beam.py

cym1 committed:
Update beam.py

jlumpe commented:
What's the status of this? I'm very interested in updates to the codegen module and would like to help if needed.

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah commented:
This should in principle be ready to go. It's a big PR so I guess it's a challenge to review.
The branch got some positive feedback here: https://github.com/bjodah/sympy/issues/6#issuecomment-391083983
It's important we get things "right" since changing things is much harder once it's in a release. If you want to review the code that would be much appreciated. Feel free to open issues at my fork (https://github.com/bjodah/sympy/issues) if you think some parts needs better explaining etc.

jlumpe committed:
Account for free symbols in codegen.ast.CodeBlock

jlumpe committed:
Check assignment types in CodeBlock.topological_sort

bjodah committed:
Merge pull request #9 from jlumpe/codeblock-free-symbols

Account for free symbols in codegen.ast.CodeBlock

jlumpe committed:
Fix test for CodeBlock.topological_sort

jlumpe committed:
Avoid conflicts between exisiting and auto-generated symbols in CodeBlock.cse()

bjodah committed:
Merge pull request #12 from jlumpe/codeblock-cse-unique-symbols

 Avoid conflicts between exisiting and auto-generated symbols in CodeBlock.cse()

bjodah committed:
Merge pull request #11 from jlumpe/fix-codeblock-toposort-test

Fix test for CodeBlock.topological_sort

jlumpe committed:
Clean up ast.Token code and add docstrings

bjodah committed:
Merge pull request #14 from jlumpe/cleanup-ast-token

Cleanup ast.Token

jashan498 committed:
resolved merge conflicts

jashan498 committed:
added max_shear_force method

jashan498 committed:
added test case for max_shear_force and some comments

jashan498 committed:
max_bmoment method was added

jashan498 committed:
added test cases for max_bmoment method

jashan498 committed:
merged master

asmeurer commented:
@bjodah how do you feel about getting this in the next release? I'd like to get it out in the next couple of weeks (before scipy).

asmeurer committed:
Revert "Changed inert Integral to active integrate."

asmeurer committed:
Change the version to 1.2.1.dev

bjodah commented:
I'm +1 to merging as is right now.

cbm755 committed:
Merge pull request #14839 from asmeurer/1.2.1.dev

Change the version to 1.2.1.dev

jashan498 committed:
added whitespaces around =

arihantparsoya committed:
Merge pull request #14826 from jashan498/patch-1

Added `max_shear_force`, `max_bmoment` and `support for statically indeterminate beams

smichr committed:
Merge pull request #14837 from sympy/revert-13937-master

Revert "Changed inert Integral to active integrate."

asmeurer commented:
What about https://github.com/bjodah/sympy/pull/15?

bjodah commented:
I don't think @jlumpe would mind directing the PR against sympy/sympy. So then the question is if this PR should be included in 1.2 or not: if it is I think we should add a note somewhere a out the module having a "provisional API".

asmeurer commented on .travis.yml:
```diff
@@ -231,7 +230,10 @@ before_install:
+        CONDA_PKGS=${TEST_OPT_DEPENDENCY/sage}
+        CONDA_PKGS=${CONDA_PKGS/autowrap/libgfortran libgcc gcc cython}
+        conda create -q -n test-environment python=$TRAVIS_PYTHON_VERSION pip ${CONDA_PKGS};
+        pip install wurlitzer;  # package for capturing C level output (tests for stdin/stderr)
```
What is wurlitzer? Can we get a conda-forge package for it?

bjodah replied:
It's for testing compiled binaries of generated C code. Here's the package:
https://github.com/minrk/wurlitzer
I opened a PR against staged-recipes:
https://github.com/conda-forge/staged-recipes/pull/6195

asmeurer commented:
Can you fix the merge conflicts with 1.2? I've refactored Travis a bit (see https://github.com/sympy/sympy/pull/14747).

asmeurer commented:
I want to include this in 1.2, so that people can start using it. 

@jlumpe's PR I'm not so sure about. We can postpone it until after the release. Changing the class structure, if we decide to do so, isn't that big of an API break.

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

asmeurer commented:
@bjodah you merged the wrong branch. This needs to be merged into `1.2`, not `master`.

bjodah committed:
Merge branch '1.2' into codegen-ast-more-nodes

bjodah commented:
I missed that at first, thanks.

asmeurer commented on .travis.yml:
```diff
@@ -23,8 +23,7 @@ matrix:
         # space separated list of optional dependencies(conda packages) to install and test
-        # autowrap installs libgfortran libgcc gcc cython
-        - TEST_OPT_DEPENDENCY="numpy scipy gmpy2 matplotlib>=2.2 theano llvmlite autowrap python-symengine=0.3.* tensorflow numexpr ipython antlr-python-runtime>=4.7,<4.8 antlr>=4.7,<4.8"
+        - TEST_OPT_DEPENDENCY="numpy scipy gmpy2 matplotlib>=2.2 theano llvmlite autowrap python-symengine=0.3.* tensorflow numexpr ipython antlr-python-runtime>=4.7,<4.8 antlr>=4.7,<4.8 sage"
```
This was the wrong merge resolution. Sage tests are run in a separate environment. See https://github.com/sympy/sympy/pull/14747/files

asmeurer commented on .travis.yml:
```diff
@@ -233,6 +232,7 @@ before_install:
 
         conda info -a;
         conda create -q -n test-environment python=$TRAVIS_PYTHON_VERSION pip ${TEST_OPT_DEPENDENCY};
+        pip install wurlitzer;  # package for capturing C level output (tests for stdin/stderr)
```
If we can get the conda-forge recipe up soon, let's move this to the conda list. Otherwise, let's change it in a later PR (I'd like to get the first release candidate out today).

asmeurer committed:
Remove sage from the TEST_OPT_DEPENDENCY variable

The sage tests are run in a separate environment. See https://github.com/sympy/sympy/pull/14747.

asmeurer commented on sympy/printing/printer.py:
```diff
@@ -235,7 +235,7 @@ def doprint(self, expr):
         return self._str(self._print(expr))
 
-    def _print(self, expr, *args, **kwargs):
+    def _print(self, expr, **kwargs):
```
Can you explain this change, and the adding of `**kwargs` to the print methods? Does this break backwards compatibility for existing printer subclasses?

bjodah replied:
Perhaps `*args` should be left intact, but when does a printer receive more than one positional argument to its `_print_*` methods? If there is such a case we should add a test case for it (demonstrating its use in e.g. a user-defined subclass).

The `**kwargs` are used for indentation when printing the ast nodes, see e.g.:
- https://github.com/sympy/sympy/pull/13100/files/717e9c2501e1ebb9c2a54c3992eb52e2a4dc357b#diff-0c241bae43013739d8f2e1ee4ae01c11R280
- https://github.com/sympy/sympy/pull/13100/files/717e9c2501e1ebb9c2a54c3992eb52e2a4dc357b#diff-0c241bae43013739d8f2e1ee4ae01c11R264

asmeurer commented on sympy/utilities/_compilation/compilation.py:
```diff
@@ -0,0 +1,627 @@
+    if clean and build_dir is not None:
+        raise ValueError("Automatic removal of build_dir is only available for temporary directory.")
+    try:
+        source_files, build_dir = _write_sources_to_build_dir(sources, build_dir)
```
If the build_dir is None, does it still clean? 

If not, the `has_fortran` etc. functions above should set clean=True.

bjodah replied:
Nice catch. Adding `clean=True` throughout.

asmeurer commented:
I don't know why they don't show up on Travis (we should figure it out), but I got this test failure running this locally. 

```pytb
_________________________________________ sympy/codegen/tests/test_algorithms.py:test_newtons_method_function__fcode _________________________________________
  File "/users/aaronmeurer/documents/python/sympy/sympy/sympy/codegen/tests/test_algorithms.py", line 59, in test_newtons_method_function__fcode
    f_mod = f_module([func], 'mod_newton')
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/codegen/futils.py", line 34, in render_as_module
    fstr = printer.doprint(mod)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/codeprinter.py", line 100, in doprint
    lines = self._print(expr).splitlines()
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/printer.py", line 274, in _print
    return getattr(self, printmethod)(expr, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/fcode.py", line 637, in _print_Module
    ).format(**mod.kwargs(apply=lambda arg: self._print(arg, **kwargs)))
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/codegen/ast.py", line 318, in kwargs
    return {k: apply(v) for k, v in kwargs.items()}
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/codegen/ast.py", line 318, in <dictcomp>
    return {k: apply(v) for k, v in kwargs.items()}
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/fcode.py", line 637, in <lambda>
    ).format(**mod.kwargs(apply=lambda arg: self._print(arg, **kwargs)))
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/printer.py", line 274, in _print
    return getattr(self, printmethod)(expr, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/codeprinter.py", line 283, in _print_CodeBlock
    return '\n'.join([self._print(i, **kwargs) for i in expr.args])
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/codeprinter.py", line 283, in <listcomp>
    return '\n'.join([self._print(i, **kwargs) for i in expr.args])
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/printer.py", line 274, in _print
    return getattr(self, printmethod)(expr, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/fcode.py", line 718, in _print_FunctionDefinition
    body=self._print(fd.body, result_name=fd.name, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/printer.py", line 274, in _print
    return getattr(self, printmethod)(expr, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/codeprinter.py", line 283, in _print_CodeBlock
    return '\n'.join([self._print(i, **kwargs) for i in expr.args])
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/codeprinter.py", line 283, in <listcomp>
    return '\n'.join([self._print(i, **kwargs) for i in expr.args])
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/printer.py", line 274, in _print
    return getattr(self, printmethod)(expr, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/codeprinter.py", line 283, in _print_CodeBlock
    return '\n'.join([self._print(i, **kwargs) for i in expr.args])
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/codeprinter.py", line 283, in <listcomp>
    return '\n'.join([self._print(i, **kwargs) for i in expr.args])
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/printer.py", line 274, in _print
    return getattr(self, printmethod)(expr, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/codeprinter.py", line 348, in _print_Statement
    return self._get_statement(self._print(arg, **kwargs))
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/printer.py", line 274, in _print
    return getattr(self, printmethod)(expr, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/fcode.py", line 458, in _print_Declaration
    s=self._print(var.symbol, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/printing/printer.py", line 274, in _print
    return getattr(self, printmethod)(expr, **kwargs)
TypeError: _print_Symbol() got an unexpected keyword argument 'result_name'
```

bjodah commented:
I can reproduce the test failure locally. Will see if I can trigger the failure on Travis before fixing it.

bjodah committed:
Clean-up after test compilation

bjodah committed:
Add cython & wurlitzer to opt. conda deps.

bjodah committed:
Merge branch '1.2' into codegen-ast-more-nodes

bjodah commented:
The failures on Travis with TEST_OPT_DEPENDENCY set look confusing, any idea what's going on there?

bjodah commented:
Here's the fix we can apply once we can trigger the test failure on Travis
```diff
diff --git a/sympy/printing/fcode.py b/sympy/printing/fcode.py
index e0a9fb1..20f5da8 100644
--- a/sympy/printing/fcode.py
+++ b/sympy/printing/fcode.py
@@ -142,7 +142,7 @@ def _lead(self):
         else:
             raise ValueError("Unknown source format: %s" % self._settings['source_format'])
 
-    def _print_Symbol(self, expr):
+    def _print_Symbol(self, expr, **kwargs):
         if self._settings['name_mangling'] == True:
             if expr not in self.mangled_symbols:
                 name = expr.name
```

asmeurer commented:
It couldn't find wurlitzer. Did the feedstock get build?

asmeurer commented:
Won't requiring kwargs on all the print methods break compatibility with existing printer subclasses?

bjodah commented:
Only when printing new classes in `.codegen.ast`. Or am I mistaken?

asmeurer commented on sympy/codegen/tests/test_algorithms.py:
```diff
@@ -0,0 +1,128 @@
+from sympy.utilities.pytest import skip, USE_PYTEST, raises
+
+cython = import_module('cython')
+wurlitzer = import_module('wurlitzer')
```
Looks like the test is skipped on travis because this file requires Cython and wurlitzer.

bjodah replied:
I see, the feedstock is empty still: https://github.com/conda-forge/wurlitzer-feedstock

bjodah replied:
Should I pip install wurlitzer in the mean time or should we wait?

asmeurer commented:
Does it pass the kwargs though to the expression printers?

asmeurer commented:
You might also need to add "codegen" to the optional dependency test list in bin/test_travis.sh

bjodah commented:
Yes. Printing of nodes is context dependent (indentation level, names of return values etc.)
An alternative to passing kwargs would be to spawn copies of the printer with modified settings in the `_print_*` methods--more complicated if you ask me.

asmeurer commented:
We should probably have set -e on the before_install in travis.yml to make errors like this fail the build earlier.

asmeurer commented:
Though I vaguely remember that not working for some reason. But that could have been a different project.

bjodah committed:
CI: pip install wurlitzer, set -e, add codegen to test list

bjodah commented:
The change of using **kwargs throughout is pretty intrusive yes. I can try to make it work with context managers. I also used it in `FCodePrinter._print_Return` (and perhaps somewhere else), so I need to figure out how to change that.

Is there an existing indentation context manager which I can plug into?

bjodah commented:
Actually, I think having a context manager temporarily overriding key/values in `printer._settings` should do the trick.

bjodah commented:
Does this look alright?
```python
@contextmanager
def temporary_printer_setting(printer, **kwargs):
    original = printer._settings.copy()
    try:
        printer._settings.update(kwargs)
        yield
    finally:
        printer._settings = original
```
(haven't tried it yet)

asmeurer commented:
I just mean like

```
@contextmanager
def indent(self):
    self.indentation_level += 1
    yield
    self.indentation_level -= 1
```

asmeurer commented:
You don't need the try/finally. That is built into the contextmanager decorator.

asmeurer commented:
You could also have a separate `printer.context` that only holds those settings that are intended to be mutated contextually, if having the ability to mutate any settings feels too off.

asmeurer commented:
I'm not really familiar enough with Fortran. Why do you need to store the return value?

asmeurer commented:
Actually I guess I'm wrong about the try/finally bit. I always though it did that automatically, but it seems it doesn't. 

```py
>>> from contextlib import contextmanager
>>> @contextmanager
... def test():
...     print(1)
...     yield
...     print(2)
>>> with test():
...     raise Exception
1
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
Exception
>>>
```

asmeurer commented on sympy/utilities/_compilation/util.py:
```diff
@@ -0,0 +1,270 @@
+    pass
+
+
+class FileNotFoundError(IOError):
```
This masks the built in FileNotFoundError in Python 3.

bjodah replied:
Thanks, will use the built-in in Py3

bjodah commented:
Will use `printer._context` (I'll initialize it in CodePrinter.__init__)

> I'm not really familiar enough with Fortran. Why do you need to store the return value?

Functions in fortran returns by assigning to the name of the function by default. The user may override the name of the return variable though.

bjodah committed:
Remove **kwargs from Printer._print_Class

bjodah commented:
Removed `**kwargs` throughout from `def _print_Class(expr, **kwargs)`.

asmeurer commented:
Great. Let's merge it as soon as the tests pass. This is the last thing blocking the release candidate. 

Re: the wurlitzer thing. It occurs to me that conda-forge might be backed up right now because of the Python 3.7 release. So let's just make sure to switch that over once it gets built. 

I hope that Python 3.7 packages can get built on conda-forge over the next week, so that I can properly test the final release against it.

asmeurer commented:
Looks like there are some test failures.

bjodah commented:
Thanks, I'm on it (looks like I forgot to run doctests locally).

bjodah committed:
Fix doctest of indented printing

bjodah committed:
revert changes to .printing.theanocode

asmeurer commented:
More test failures.

bjodah committed:
Use TemporaryDirectory contex manager in tests

bjodah commented:
Deprecation warnings about `imp`:
```
DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
```
Could we ignore those somehow? (I seem to remember that the version of importlib shipped with py27 lacks too many features to be a useful replacement).

asmeurer commented:
Can you use `imp` only in Python 2? That way when we drop Python 2 support it will also drop the usage of imp.

bjodah committed:
Use importlib in Py3

bjodah commented:
Gave it a shot. (tests pass locally under py27 & py35). Going to bed now so won't be able to fix any test failures immediately. Feel free to push directly if anything minor is still missing.

asmeurer commented:
I don't know how to fix the latest errors.

```
________________ sympy/codegen/tests/test_fnodes.py:test_bind_C ________________
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.2rc1-py3.6.egg/sympy/utilities/runtests.py", line 1259, in _timeout
    function()
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.2rc1-py3.6.egg/sympy/codegen/tests/test_fnodes.py", line 200, in test_bind_C
    ], build_dir=folder)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.2rc1-py3.6.egg/sympy/utilities/_compilation/compilation.py", line 578, in compile_link_import_strings
    mod = compile_link_import_py_ext(source_files, build_dir=build_dir, **kwargs)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.2rc1-py3.6.egg/sympy/utilities/_compilation/compilation.py", line 527, in compile_link_import_py_ext
    cplus=any_cplus_src(sources), **link_kwargs)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.2rc1-py3.6.egg/sympy/utilities/_compilation/compilation.py", line 248, in link_py_so
    libraries=libraries, library_dirs=library_dirs, **kwargs)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.2rc1-py3.6.egg/sympy/utilities/_compilation/compilation.py", line 188, in link
    runner.run()
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.2rc1-py3.6.egg/sympy/utilities/_compilation/runners.py", line 193, in run
    raise CompileError(msg)
distutils.errors.CompileError: Error executing '/usr/bin/gfortran -pthread -shared -o /tmp/tmpjchf3cva/_rms.cpython-36m-x86_64-linux-gnu.so ./rms.o ./_rms.o -lpython3.6m' in /tmp/tmpjchf3cva (exited status 1):
 /usr/bin/ld: cannot find -lpython3.6m
collect2: error: ld returned 1 exit status
```

bjodah committed:
Set compiler environment variables for conda env

bjodah committed:
wurlitzer from conda-forge

bjodah commented:
Needed to set LD_LIBRARY_PATH.
Tests are now passing.

bjodah commented:
Looks like test_fnodes.py has a bunch of skipped tests though:
```
sympy/codegen/tests/test_fnodes.py[12] .sssss.....s
```
due to `has_fortran` returning `False`. Not sure why since `gfortran` should be installed.

bjodah committed:
CI: Require compilers to work in specific Travis runs

bjodah committed:
CI: set LD_LIBRARY_PATH

bjodah committed:
More specific exception for missing compiler

asmeurer commented:
I'm not sure about setting LD_LIBRARY_PATH. It's generally only for debugging or development. It shouldn't be required for sympy to work. Will this work outside of Travis when it isn't set?

asmeurer commented:
I get some test failures on my Mac

```
_________________________________________ sympy/codegen/tests/test_algorithms.py:test_newtons_method_function__ccode _________________________________________
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/codegen/tests/test_algorithms.py", line 44, in test_newtons_method_function__ccode
    ], build_dir=folder, compile_kwargs=compile_kw)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/_compilation/compilation.py", line 578, in compile_link_import_strings
    mod = compile_link_import_py_ext(source_files, build_dir=build_dir, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/_compilation/compilation.py", line 525, in compile_link_import_py_ext
    cwd=build_dir, **compile_kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/_compilation/compilation.py", line 97, in compile_sources
    dstpaths.append(src2obj(f, Runner, cwd=cwd, **file_kwargs))
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/_compilation/compilation.py", line 358, in src2obj
    **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/_compilation/compilation.py", line 437, in pyx2obj
    interm_c_file = simple_cythonize(pyxpath, destdir=destdir, cwd=cwd, **cy_kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/_compilation/compilation.py", line 295, in simple_cythonize
    shutil.move(os.path.join(os.path.dirname(src), c_name), destdir)
  File "/Users/aaronmeurer/anaconda3/lib/python3.5/shutil.py", line 558, in move
    copy_function(src, real_dst)
  File "/Users/aaronmeurer/anaconda3/lib/python3.5/shutil.py", line 257, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/Users/aaronmeurer/anaconda3/lib/python3.5/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '/var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/tmpy9g_47kd/_newton.c'
```

bjodah commented:
Strange, but not surprising since I haven't tested it on macOS. Windows will probably not work either. I did however (naively) expect the "has_fortran" and "has_c" functions to catch any problems.

Should we add conditional `SKIP`s to the test cases based on `sys.platform.lower() == 'darwin'` and `os.name == 'nt'`? (I simply don't really have the time necessary to fix this right now -- things like [scipy2017codegen/template.pyxbld](https://github.com/sympy/scipy-2017-codegen-tutorial/blob/85382bfed316d2dde7cfdc83cef20a255efe2c2b/scipy2017codegen/template.pyxbld) are probably necessary (and that took a considerable amount of time and testing to get right).

The most important thing is that the generated code gets tested, though it would be nice if that could be done locally on windows/macOS machines with compilers.

bjodah commented:
An alternative would be to drop `.util._compilation` completely until we have something "perfect", but then the generated code would have to be tested manually (something I think is worse than having it Linux/Travis only to start with).

asmeurer commented:
So the point of the file is only for testing, not for actual end-user usage?

asmeurer commented:
@bjodah can you pop onto gitter (the main sympy/sympy channel)? I think we could get this resolved faster there.

bjodah committed:
Fixes to compilation

- may_xfail on macOS & Windows
- Fix link_py_so for python built without shared library
- More fine grained skips in tests of codegen

asmeurer commented:
OK, I am +1 to merge as soon as Travis passes.
